docs/dev/ui_description.rst

```diff
@@ -0,0 +1,259 @@
+trailets on the custom widget class. The template reads and responds to
+state and state changes on the custom widget, and the custom widget need
+not know or care about how that state is being represented (i.e. the
+only state is that of the custom widget).
+
+With this approach in hand, all widgets in the glupyter GUI are composed
+of two files: the python file declaring the widget class (e.g.
+``Toolbar``), and the ``.vue`` file containing the template describing
+how the widget is to be rendered using components from the Vuetify
+library in the nominal Vue framework. In this way, there is a clear
+separation between the state of each widget (contained in the python
+file) and the view of that state (contained in the Vuetify file). Below
+is an example of what such a widget might looking like:
+
+.. code:: markup
+
+    <template>
+      <v-toolbar>
+        <v-toolbar-side-icon></v-toolbar-side-icon>
+        <v-toolbar-title>Title</v-toolbar-title>
+        <v-spacer></v-spacer>
+        <v-toolbar-items class="hidden-sm-and-down">
+          <v-btn flat>Link One</v-btn>
+          <v-btn flat>Link Two</v-btn>
+          <v-btn flat>Link Three</v-btn>
+        </v-toolbar-items>
+      </v-toolbar>
+    </template>
+
+.. code:: python
+
+    with open(os.path.join(os.path.dirname(__file__), "toolbar.vue")) as f:
+        TEMPLATE = f.read()
+
+
+    class Toolbar(VuetifyTemplate):
+        template = Unicode(TEMPLATE).tag(sync=True)
+
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            ...
+
+The state of the widget is contained in attributes on the python class
+which allows them to be referenced in the Vuetify template. Notice in
+the example below that the ``v-btn`` instances simply respond to the
+state of the ``Toolbar`` widget's ``button_names`` attribute, and the
+``Toolbar`` class could know nothing about *how* that state is being
+represented.
+
+.. code:: markup
+
+    <v-toolbar-items class="hidden-sm-and-down">
+      <v-btn flat v-for="name in button_names">{{ name }}</v-btn>
+    </v-toolbar-items>
+
+.. code:: python
+
+    class Toolbar(VuetifyTemplate):
+        template = Unicode(TEMPLATE).tag(sync=True)
+        button_names = List(['One', 'Two', 'Three']).tag(sync=True)
+
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            ...
+
+The design of the interface can be broken down into three main areas:
+the toolbar, the navigation drawer, and the dock area. Each of these
+areas represents a single primary widget in the web-based application
+built using ``ipyvuetify``, unified in the ``Application`` class.
+
+Widget communication
+--------------------
+
+There are two fundamental forms of communication between widgets:
+direction communication using the ``observer`` pattern, and global
+communication using the centralized event hub provided by Glue.
```

Comment:
Just to be clear, we might also use the callback properties in glue as a third way - this is basically very similar conceptually to traitlets.